flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericExactlyOnceSink.java

```diff
@@ -0,0 +1,195 @@
+
+	public GenericExactlyOnceSink(CheckpointCommitter committer) {
+		if (committer == null) {
+			throw new IllegalArgumentException("CheckpointCommitter argument must not be null.");
+		}
+		this.committer = committer;
+		this.id = UUID.randomUUID().toString();
+	}
+
+	@Override
+	public void open() throws Exception {
+		committer.setOperatorId(id);
+		committer.setOperatorSubtaskId(getRuntimeContext().getIndexOfThisSubtask());
+		committer.open();
+	}
+
+	public void close() throws Exception {
+		committer.close();
+	}
+
+	/**
+	 * Saves a handle in the state.
+	 *
+	 * @param checkpointId
+	 * @throws IOException
+	 */
+	private void saveHandleInState(final long checkpointId) throws IOException {
+		//only add handle if a new OperatorState was created since the last snapshot
+		if (out != null) {
+			StateHandle<DataInputView> handle = out.closeAndGetHandle();
+			state.pendingHandles.put(checkpointId, handle);
+			out = null;
+		}
+	}
+
+	@Override
+	public StreamTaskState snapshotOperatorState(final long checkpointId, final long timestamp) throws Exception {
+		StreamTaskState taskState = super.snapshotOperatorState(checkpointId, timestamp);
+		saveHandleInState(checkpointId);
+		taskState.setFunctionState(state);
+		return taskState;
+	}
+
+	@Override
+	public void restoreState(StreamTaskState state, long recoveryTimestamp) throws Exception {
+		super.restoreState(state, recoveryTimestamp);
+		this.state = (ExactlyOnceState) state.getFunctionState();
+		out = null;
+	}
+
+	@Override
+	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
+		super.notifyOfCompletedCheckpoint(checkpointId);
+
+		synchronized (state.pendingHandles) {
+			Set<Long> pastCheckpointIds = state.pendingHandles.keySet();
+			Set<Long> checkpointsToRemove = new HashSet<>();
+			for (Long pastCheckpointId : pastCheckpointIds) {
+				if (pastCheckpointId <= checkpointId) {
+					if (!committer.isCheckpointCommitted(pastCheckpointId)) {
+						StateHandle<DataInputView> handle = state.pendingHandles.get(pastCheckpointId);
+						DataInputView in = handle.getState(getUserCodeClassloader());
+						sendValue(new NonReusingMutableToRegularIteratorWrapper<>(new InputViewIterator<>(in, serializer), serializer));
+						committer.commitCheckpoint(pastCheckpointId);
+					}
+					checkpointsToRemove.add(pastCheckpointId);
+				}
+			}
+			for (Long toRemove : checkpointsToRemove) {
+				StateHandle<DataInputView> handle = state.pendingHandles.get(toRemove);
+				state.pendingHandles.remove(toRemove);
+				handle.discardState();
+			}
+		}
+	}
+
+
+	/**
+	 * Write the given element into the backend.
+	 *
+	 * @param value value to be written
+	 * @throws Exception
+	 */
+	protected abstract void sendValue(Iterable<IN> value) throws Exception;
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		IN value = element.getValue();
+		if (serializer == null) {
+			typeInfo = TypeExtractor.getForObject(value);
```

Comment:
This won't work here. Assume that you have a stream of some super class `S`. The first element you see is a sub class `A` so you will instantiate a serializer of this type. But when you see next another sub class `B`, then it will fail.

I think you should pass the serializer to the operator when it is instantiated on the client side.